rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,346 @@
+
+* Authentication tokens
+  * To submit requests to HEC in general, an authentication token must be
+    included in request headers. Indexer acknowledgement is enabled per
+    authentication token in Splunk. The `splunk_hec` source currently supports a
+    list of valid authentication tokens. The `splunk_hec` sinks currently
+    support configuring a single authentication token.
+
+* Splunk channels
+  * Upon enabling indexer acknowledgement, Splunk requires HEC requests to
+    include a channel ID (any valid `guid` value). Querying
+    `/services/collector/ack` also requires a channel ID.
+  * `ackId`’s are assigned per-channel.
+
+* Pending acks
+  * Clients are not required or guaranteed to query for `ackId` status. To avoid
+    running out of memory from `ackId` build up, Splunk offers a few
+    configuration options to limit the number of pending `ackId`’s both overall
+    and per-channel.
+  * Splunk does not explicitly indicate whether event data is dropped or still
+    pending processing. Either of these states will result in `{ack_id}: false`
+    status. Splunk only advises that after a certain amount of time (e.g. 5
+    minutes), the data can be considered dropped.
+
+## Proposal: `splunk_hec` Source Indexer Acknowledgement
+
+### User Experience
+
+Users can configure the `splunk_hec` source with additional indexer acknowledgement related settings.
+
+```{toml}
+[sources.splunk_hec]
+type = “splunk_hec”
+...
+acknowledgements.enabled = true
+acknowledgements.max_pending_acks = 1_000_000
+```
+
+* `acknowledgements.enabled`  This defaults to `false` matching Splunk HEC's
+  opt-in behavior.
+* `acknowledgements.max_pending_acks` With acknowledgements enabled, this
+  controls the maximum number of pending query ackId's (to avoid memory issues).
+  This defaults to `10_000_000` matching Splunk HEC's default.
+
+Since Vector does not share Splunk’s internal constraints, we can relax certain
+protocol requirements to avoid unnecessary complexity. Specifically,
+
+* Authentication tokens
+  * Enabling indexer acknowledgements will be an overall `splunk_hec` source
+    configuration rather than a per-token configuration. Users can configure a
+    secondary Splunk source without acknowledgements if necessary, and/or users
+    can ignore the `ackId`s for requests that do not participate in
+    acknowledgement.
+
+* Splunk Channels
+  * The `splunk_hec` source will not require channel IDs for acknowledgement.
+    Currently, we store channel IDs as an additional `LogEvent` field. If users
+    intend to move data from a `splunk_hec` source to a `splunk_hec` sink,
+    passing through the channel IDs can be helpful.
+  * Rather than assign `ackId`’s per channel, the `splunk_hec` source will
+    assign `ackId`’s across all its requests. In other words, if the source
+    receives two requests with differing channel IDs, it will reply with `ackId
+    = 0` and `ackId = 1` rather than `ackId = 0` and `ackId = 0`.
```

Comment:
Looking at the Splunk docs, they mention that they use channel identifiers to keep clients separate and avoid, for example, a fast client starving a slow client. We don't have to deal with that particular issue (right now at least), but I could see the case where a very fast client could saturate the acks and cause slower clients to have their acks be evicted before they can query them. I'm imagining a scenario where:

* Client A using channel C1 sends a request, and gets an ack
* Client B using channel C2 sends more requests than the allowed maximum pending acks
* Client A's ack gets evicted

Given that, and the fact that Splunk users are likely to be used to configuring the limits at the channel level, I'm wondering if it wouldn't make more sense to more closely match the Splunk indexer configuration options. That is: allow the configuration of multiple channels, each with its own pending ack limit.

Reply:
Ah, that is a great point. In that case, it does seem like replicating Splunk's channel requirement is worthwhile. I'll update the writing unless there are other thoughts here. I'd be curious to know how common the scenario above might be but handling channels seems like the more robust option.

Reply:
Yeah, I think one other strong motivator for doing it is that users in the Splunk ecosystem will already be familiar this sort of configuration.

Reply:
It also, I think, makes vector more drop-in compatible if we mimic their configuration.

Reply:
s/mimic/take inspiration from/ :)

Reply:
Updated with [channel configuration](https://github.com/vectordotdev/vector/blob/rfc-indexer-acknowledgements/rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md#proposal-splunk_hec-source-indexer-acknowledgement) and [channel behavior](https://github.com/vectordotdev/vector/blob/rfc-indexer-acknowledgements/rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md#channel-behavior)